### ISSUE TITLE
Unexpected behavior in model batch prediction?

### ISSUE THREAD

**Issue by Christianfoley** [mehta-lab/microDL#225]
I've been noticing that when I perform batch prediction (pass in tensors with higher than required number of slices to the model) I am seeing artifacts in the predicted slices of the batches that I don't see in the non-batched prediction.

I can reproduce these artifacts with the following script:
```python
# %%
import numpy as np
import torch
import matplotlib.pyplot as plt
import sys

sys.path.insert(0, "/home/christian.foley/virtual_staining/workspaces/microDL") #change to your directory

import micro_dl.utils.aux_utils as aux_utils
import micro_dl.inference.inference as torch_inference_utils

# %%
# Use the config to initialize a predictor. The predictor just makes it easy to
# initialize the model with the parameters from its training config
config = (
    "/hpc/projects/CompMicro/projects/virtualstaining/"
    "torch_microDL/config_files/2022_HEK_nuc_mem_Soorya/TestData_HEK_2022_04_16/"
    "inference_config_processed_hek_no_perturb_256_256.yml"
)

config = aux_utils.read_config(config)
device = torch.device(config.get("device"))

# Initialize and run a predictor
torch_predictor = torch_inference_utils.TorchPredictor(
    config=config,
    device=device,
)
torch_predictor.load_model()

# %%
# Extract our relevant objects (dataset object will act as a library for convenience)
dataset = torch_predictor.dataset
plate = dataset.data_plate
model = torch_predictor.model


# %%
# Get a chunk of data and normalize it
phase_index = plate.channel_names.index("Phase3D")
slice_chunk = plate["plate"][0][0].data[0, phase_index, 55:60, ...]
slice_batch = plate["plate"][0][0].data[0, phase_index, 53:62, ...]
print(slice_chunk.shape, slice_batch.shape)  # (5, 2048, 2048) (9, 2048, 2048)

norm_stats = dataset._get_normalization_statistics("Phase3D")
normed_chunk = np.expand_dims(dataset._normalize(slice_chunk, norm_stats), (0, 1))
normed_batch = np.expand_dims(dataset._normalize(slice_batch, norm_stats), (0, 1))
print(normed_chunk.shape, normed_batch.shape)  # (1, 1, 5, 2048, 2048) (1, 1, 9, 2048, 2048)

# %%
# Run through model (2.5d model only accepts/returns 5d inputs)
slice_pred = model(torch.tensor(normed_chunk, device="cuda")).detach().cpu().numpy()
batch_pred = model(torch.tensor(normed_batch, device="cuda")).detach().cpu().numpy()
print(slice_pred.shape, batch_pred.shape)  # (1, 2048, 2048) (5, 2048, 2048)

# %%
# Plot to show artifacts
plt.figure(figsize=(3, 3))
plt.title("center slice prediction of a 5-slice input")
plt.imshow(slice_pred[0, 0, 0], vmax=35)
plt.show()

fig, ax = plt.subplots(1, 5, figsize=(15, 3))
plt.suptitle("5 center slices predictions from of a 9-slice input")
for i in range(5):
    ax[i].imshow(batch_pred[0, 0, i], vmax=35)
plt.show()

plt.figure(figsize=(3, 3))
plt.title("difference between center slices predictions\n of both inputs")  # this should be 0
plt.imshow(slice_pred[0, 0, 0] - batch_pred[0, 0, i], vmax=35)
plt.show()
````
**First plot:**
![Screen Shot 2023-04-20 at 8 06 27 AM](https://user-images.githubusercontent.com/59987950/233408210-099af174-651f-445a-88fa-d36fb10dafc7.png)

**Second plot:**
![Screen Shot 2023-04-20 at 8 07 17 AM](https://user-images.githubusercontent.com/59987950/233408458-b78e7c6f-4eba-4587-879f-3f3d09087fb2.png)

**Third plot (comparing first plot and middle-slice of 2nd plot):**
![Screen Shot 2023-04-20 at 8 07 33 AM](https://user-images.githubusercontent.com/59987950/233408527-55a82b72-5cbb-4508-b7bf-1a6245d8493f.png)

The model skip connection interruptions are the only place z-axis convolution is implemented [conv block here](https://github.com/mehta-lab/microDL/blob/40dcd0e026af9f468e2be9ed11ffa0ffc0ef90df/micro_dl/torch_unet/networks/Unet25D.py#L176-L185) and [skip connection kernel here](https://github.com/mehta-lab/microDL/blob/40dcd0e026af9f468e2be9ed11ffa0ffc0ef90df/micro_dl/torch_unet/networks/Unet25D.py#L168). We should see that the center slice is the same because of the sliding window. Any ideas what is going on here @mattersoflight @ziw-liu?

**Comment by ziw-liu** [comment]
> The model skip connection interruptions are the only place z-axis convolution is implemented [conv block here](https://github.com/mehta-lab/microDL/blob/40dcd0e026af9f468e2be9ed11ffa0ffc0ef90df/micro_dl/torch_unet/networks/Unet25D.py#L176-L185) and [skip connection kernel here](https://github.com/mehta-lab/microDL/blob/40dcd0e026af9f468e2be9ed11ffa0ffc0ef90df/micro_dl/torch_unet/networks/Unet25D.py#L168). We should see that the center slice is the same because of the sliding window.

But the encoder layers are all 3D. So I'm guessing it's due to (the lack of) Z padding in 3D convolution, as the bottom and top slices of the 9-slice output has less noise.

**Comment by ziw-liu** [comment]
Also I'm a bit confused about the definition of the 'batch prediction' here. The number of Z-slices is not the batch size right?

**Comment by Christianfoley** [comment]
> But the encoder layers are all 3D
The encoder layer kernels are 1 in the Z dimension.

**Comment by ziw-liu** [comment]
Don't know if it has any effect, but both your script and `TorchPredictor.predict_image()` doesn't put the model in inference mode.

**Comment by Christianfoley** [comment]
Good catch, let me change that! It shouldn't be the cause of this issue since I think it would only affect dropout here.

**Comment by ziw-liu** [comment]
> The encoder layer kernels are 1 in the Z dimension.

As we discovered in the meeting, the encoder layers does perform 3D convolutions with kernel size of 3 along Z:

https://github.com/mehta-lab/microDL/blob/40dcd0e026af9f468e2be9ed11ffa0ffc0ef90df/micro_dl/torch_unet/networks/Unet25D.py#L118

Since the effective receptive field along Z is large wrt the Z input size during training, the layers will take in very different values when the input size changes.

In practice we can still use these models, and parallelize by using the same Z shape and stacking on the batch dimension.

**Comment by Christianfoley** [comment]
Resolved with commit [https://github.com/mehta-lab/microDL/commit/d80516c379e81fdffac4a44a3bfb16d9380a0544](https://github.com/mehta-lab/microDL/commit/d80516c379e81fdffac4a44a3bfb16d9380a0544).
